src/json/mod.rs

```diff
@@ -24,63 +31,88 @@ pub fn create_json(host: &AnalysisHost, crate_name: &str) -> Result<String> {
 
     let root_def = host.get_def(root_id)?;
 
-    fn recur(id: &analysis::Id, host: &AnalysisHost) -> Vec<analysis::Def> {
-        let mut ids = Vec::new();
-        let mut defs = host.for_each_child_def(*id, |id, def| {
-            ids.push(id);
-            def.clone()
-        }).unwrap();
-
-        let child_defs: Vec<analysis::Def> = ids.into_par_iter()
-            .map(|id: analysis::Id| recur(&id, host))
-            .reduce(Vec::default, |mut a: Vec<analysis::Def>,
-             b: Vec<analysis::Def>| {
-                a.extend(b);
-                a
-            });
-        defs.extend(child_defs);
-        defs
-    }
+    // create the main `Document`
+    let mut document = Document::new()
+        .ty(String::from("crate"))
+        .id(crate_name.to_string())
+        .attributes(String::from("docs"), root_def.docs);
 
-    let mut included: Vec<Document> = Vec::new();
-    let mut relationships: HashMap<String, Vec<Data>> = HashMap::with_capacity(METADATA_SIZE);
+    // Now that we have that, it's time to get the children; these are
+    // the top-level items for the crate.
+    let ids = host.for_each_child_def(root_id, |id, _def| id).unwrap();
+
+
+    // Now, we push all of those children onto a channel. The channel functions
+    // as a work queue; we take an item off, process it, and then if it has
+    // children, push them onto the queue. When the queue is empty, we've processed
+    // everything.
+    //
+    // Additionally, we generate relationships between the crate itself and
+    // these ids, as they're at the top level and hence linked with the crate.
 
-    for def in recur(&root_id, host) {
-        let (ty, relations_key) = match def.kind {
+    let (sender, receiver) = mpsc::channel();
+
+    for id in ids {
+        sender.send(id).unwrap();
+
+        let def = host.get_def(id).unwrap();
+
+        let (ty, child_ty) = match def.kind {
             DefKind::Mod => (String::from("module"), String::from("modules")),
             DefKind::Struct => (String::from("struct"), String::from("structs")),
             _ => continue,
         };
 
+        let data = Data::new().ty(ty.clone()).id(def.qualname.clone());
+
+        document.add_relationship(child_ty, data);
+    }
+
+    // the loop below is basically creating this vector
+    let mut included: Vec<Document> = Vec::new();
+
+    while let Ok(id) = receiver.try_recv() {
+        // push each child to be processed itself, and also record
+        // their ids so we can create the relationships for later
+        let child_ids = host.for_each_child_def(id, |id, _def| {
+            sender.send(id).unwrap();
+            id
+        })?;
+
+        // process this one
+
+        // question: we could do this by cloning it in the call to for_each_child_def
```

Comment:
s/question/Question

As for the answer we'd need some kind of benchmark. I think we don't need to worry about this too much for now and should focus more on getting things working then optimization.